crates/environ/src/module.rs

```diff
@@ -259,6 +259,100 @@ impl ModuleTranslation<'_> {
+        const MAX_FUNC_TABLE_SIZE: u32 = 1024 * 1024;
+
+        let num_imported_tables = self.module.num_imported_tables;
+        for (table_index, init) in &mut self.module.table_initialization {
+            // If this is an imported table, we can't provide a
+            // preconstructed table for it, because its values depend
+            // on the imported table overlaid with whatever segments
+            // we have.
+            if table_index.index() < num_imported_tables {
+                continue;
+            }
+
+            // If this is not a funcref table, then we can't support a
+            // pre-computed table of function indices.
+            if self.module.table_plans[table_index].table.wasm_ty != WasmType::FuncRef {
+                continue;
+            }
+
+            // If there is no initial content for the table, then
+            // there is nothing to do here.
+            let init = match init {
+                Some(init) => init,
+                None => continue,
+            };
+
+            // Get the list of segment initializers. If we're in some
+            // other mode, then nothing to do here.
+            let segments = match init.as_segments() {
+                Some(segments) => segments,
+                None => continue,
+            };
+
+            // Check all segments to see if any have a dynamic base
+            // index. If any are, then we can't build an array of
+            // initial contents that is independent of instantiation
+            // state (values of globals), so remain in Segments mode.
+            let has_dynamically_placed_segment =
+                segments.iter().any(|segment| segment.base.is_some());
+            if has_dynamically_placed_segment {
+                continue;
+            }
+
+            // Check the table size. If too large to justify a dense
+            // data structure (array with entries for all elements),
+            // then remain in Segments mode.
+            let table_size = self.module.table_plans[table_index].table.minimum;
+            if table_size > MAX_FUNC_TABLE_SIZE {
+                continue;
+            }
+
+            // Check all segments to see if any are out-of-bounds. If
+            // any are, then we fall back to the default code-paths to
+            // handle the error and get the final state just before
+            // the trap correct.
+            let has_out_of_bounds = segments.iter().any(|segment| {
+                segment
+                    .offset
+                    .checked_add(segment.elements.len() as u32)
+                    .unwrap_or(u32::MAX)
+                    > table_size
+            });
+            if has_out_of_bounds {
+                continue;
+            }
+
+            // Now we can commit to building an array of FuncIndex
+            // values.
+            let mut elements: Vec<Option<FuncIndex>> = vec![None; table_size as usize];
```

Comment:
It might be worth double-checking but I think that `Option<FuncIndex>` is actually a 64-bit value whereas `FuncIndex` is a 32-bit value. To make this more compact I think we could represent `None` with `FuncIndex::reserved_value()`.

Personally I'm not a fan of one value of `FuncIndex` meaning "none" but it's the design we already have and changing that is beyond the scope of this PR. Despite my personal thoughts, though, it does lend itself well to this use case.

Reply:
Done!